_episodes/l2_02_basic_unit_testing.md

```diff
@@ -0,0 +1,293 @@
+
+1. **Setup**: In the setup phase, we prepare the necessary preconditions for the unit test. This involves creating any required objects, initialising variables, and setting up the environment to mimic the desired test scenario. The setup step ensures that the unit being tested has the necessary dependencies and context to execute successfully.
+2. **Execution**: The execution step involves invoking the unit under test with the specified inputs or parameters. This is the actual execution of the code being tested. The unit is executed with the predetermined inputs, and its output or behavior is observed.
+3. **Verification**: In the verification step, we check whether the actual output or behaviour matches the expected result. This typically involves making assertions or comparisons between the observed output and the expected output. If the assertions pass, the test is considered successful. Otherwise, if any assertion fails, it indicates a discrepancy between the expected and actual outcomes, highlighting a potential issue in the unit being tested.
+
+This 3 step structure if often referred as `Arrange-Act-Assert (AAA)` in some textbooks and online resources. 
+
+
+<br> </br>
+## 6. Writing Effective Unit Tests
+
+For writing your Unit Tests, it is advisable to follow the guidelines given below.
+
+1. **Clear and Descriptive Names**: Choose meaningful names for your test methods that accurately describe the scenario being tested. A good test method name should clearly convey the input, expected behavior, or outcome being verified. Avoid vague or generic names that don't provide sufficient information about the purpose of the test.
+
+   a. `calculateTotal_WithValidInputs_ShouldReturnCorrectSum`
+
+   b. `validateEmail_WithInvalidFormat_ShouldReturnFalse`
+
+   
+
+2. **Consistent Formatting**: Consistency in naming conventions helps maintain a uniform and predictable test suite. Choose a naming style and stick to it throughout your test methods. Some common conventions include using **CamelCase** or **underscore-separated** words. Additionally, consider using a prefix like `test_` or a suffix like `should` to distinguish test methods from regular code.
+
+   
+
+3. **Single Responsibility**: Each test method should focus on testing a single aspect or behaviour of the unit under test. Avoid testing multiple scenarios within a single test method, as it can make the test less readable and harder to diagnose when failures occur. Instead, break down complex scenarios into multiple smaller tests, each targeting a specific case or condition.
+
+   
+
+4. **Test Data Organization**: Separate the test data from the test methods. Consider using dedicated variables or data structures to store test data. This allows for better readability and maintainability, as changes to the test data can be easily managed without modifying the test methods themselves.
+
+
+
+## 7. Unit Test for the Fibonacci Sequence
```

Comment:
I think this section should come before the theory, actually. The reason is that it's often easier to understand with a concrete examples and we want to make sure the students really understand what a unit test is and what it's for before delving into more abstract topics.

Reply:
I like to learn things from bottom up rather than top down approach. This is why the elements (to build a test program and its steps) are introduced first and then a full example.

Reply:
I agree it might be better before the theory, but let's leave it as it is, for now. This is just the first version and, for sure, we will need to update the course in the future once we have delivered it once of twice and identify what works and what doesn't.